server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -347,10 +346,9 @@ public ClusterState execute(ClusterState currentState) {
             if (found == false) {
                 repositoriesMetadata.add(new RepositoryMetadata(request.name(), request.type(), request.settings()));
             }
-            repositories = new RepositoriesMetadata(repositoriesMetadata);
-            mdBuilder.putDefaultProjectCustom(RepositoriesMetadata.TYPE, repositories);
             changed = true;
-            return ClusterState.builder(currentState).metadata(mdBuilder).build();
+            var updatedRepositories = new RepositoriesMetadata(repositoriesMetadata);
+            return currentState.copyAndUpdateProject(project.id(), b -> b.putCustom(RepositoriesMetadata.TYPE, updatedRepositories));
```

Comment:
I wonder whether we should have this as
```java
return ClusterState.builder(currentState).putProjectMetadata(
    ProjectMetadata.builder(project).putCustom(RepositoriesMetadata.TYPE, repositories)
).build();
```
It is not much more verbose and avoids creation of an lambda object. As explained above, I prefer it also because it helps maintain the existing code structure as much as possible.